compiler/qsc_formatter/src/formatter.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use qsc_data_structures::span::Span;
+use qsc_frontend::lex::{
+    concrete::{self, ConcreteToken, ConcreteTokenKind},
+    cooked::TokenKind,
+    Delim,
+};
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Debug)]
+pub struct Edit {
+    pub span: Span,
+    pub new_text: String,
+}
+
+impl Edit {
+    fn new(lo: u32, hi: u32, new_text: &str) -> Self {
+        Self {
+            span: Span { lo, hi },
+            new_text: new_text.to_string(),
+        }
+    }
+}
+
+fn make_indent_string(level: usize) -> String {
+    "    ".repeat(level)
+}
+
+/// Applies formatting rules to the given code str, generating edits where
+/// the source code needs to be changed to comply with the format rules.
+pub fn format(code: &str) -> Vec<Edit> {
```

Comment:
This API seems very limited to supporting the language service. For testing, and other uses, it would seem handy to have `pub fn format(code: &str) -> String` with the formatted output and have this function be `pub fn calculate_edits(code: &str) -> Vec<Edit>`.

Reply:
That is a nice suggestion and made the tests much more readable. Thanks!